modules/apps/frontend-js/frontend-js-react-web/src/main/resources/META-INF/resources/js/A11y.es.js

```diff
@@ -0,0 +1,272 @@
+											: alert.impact === 'moderate'
+											? 'warning-full'
+											: 'exclamation-full'
+									}
+								/>
+							</div>
+							<div className="autofit-col">
+								<a
+									className="a11y-help"
+									href={alert.helpUrl}
+									rel="noopener noreferrer"
+									target="_blank"
+								>
+									{alert.help}
+								</a>
+								<span className="text-muted">
+									Fix {all.length > 0 ? 'all' : 'any'} of the
+									following
+								</span>
+							</div>
+						</div>
+						{all.length > 0 && (
+							<ul className="a11y-list">
+								{all.map((check) => (
+									<li key={check.id}>{check.message}</li>
+								))}
+							</ul>
+						)}
+						{any.length > 0 && (
+							<ul className="a11y-list">
+								{any.map((check) => (
+									<li key={check.id}>{check.message}</li>
+								))}
+							</ul>
+						)}
+					</article>
+				);
+			})}
+		</ClayPopover>
+	);
+}
+
+export default function A11y({children, enable = true}) {
+	const [alerts, setAlerts] = useState([]);
+	const [disabled, setDisabled] = useState(!enable);
+	const workInProgressRef = useRef(null);
+	const idleIdRef = useRef(null);
+	const childrenRef = useRef(null);
+
+	const alertsByNode = useMemo(() => segmentAlertsByNode(alerts), [alerts]);
+
+	useEffect(() => {
+		if (!childrenRef.current || disabled) {
+			return;
+		}
+
+		if (idleIdRef.current && window.cancelIdleCallback) {
+			window.cancelIdleCallback(idleIdRef.current);
+			idleIdRef.current = null;
+		}
+
+		const getAlerts = () =>
+			axe
+				.run(childrenRef.current, {reporter: 'v2'})
+				.then((results) => setAlerts(results.violations))
+				.catch(console.error);
+
+		const workCallback = () => {
+			if (!workInProgressRef.current) {
+				workInProgressRef.current = getAlerts()
+					.then(() => {
+						workInProgressRef.current = null;
+					})
+					.catch(() => {
+						workInProgressRef.current = null;
+					});
+			}
+		};
+
+		const mutationCallback = () => {
+			if (window.requestIdleCallback) {
+				if (idleIdRef.current) {
+					window.cancelIdleCallback(idleIdRef.current);
+				}
+
+				idleIdRef.current = window.requestIdleCallback(workCallback);
+			}
+			else {
+				workCallback();
+			}
+		};
+
+		const observer = new MutationObserver(mutationCallback);
+
+		observer.observe(childrenRef.current, {
+			attributes: true,
+			childList: true,
+			subtree: true,
+		});
+
+		return () => {
+			if (idleIdRef.current && window.cancelIdleCallback) {
+				window.cancelIdleCallback(idleIdRef.current);
+				idleIdRef.current = null;
+				workInProgressRef.current = null;
+			}
+			observer.disconnect();
+		};
+	}, [disabled]);
+
+	useEffect(() => {
+		const handler = (event) => {
+			if (event.ctrlKey && event.key === 'd' && enable) {
+				setDisabled(!disabled);
+			}
+		};
+
+		document.addEventListener('keydown', handler);
+
+		return () => {
+			document.removeEventListener('keydown', handler);
+		};
+	}, [disabled, enable]);
```

Comment:
You might be able to replace some of this with [`useEventListener()`](https://github.com/liferay/liferay-portal/blob/4aa21b17b64419f2fdf9cd42bfa4612a8e4586d1/modules/apps/frontend-js/frontend-js-react-web/src/main/resources/META-INF/resources/js/hooks/useEventListener.es.js).

Reply:
What do you think about creating a hook on frontend-js-web called `useHotkeys`?

use case:


```jsx
import KEYS, {useHotKeys} from 'frontend-js-web';

function myComponent(...props) {
    useHotKeys(KEYS.W, () => 'w was pressed');
    useHotkeys([KEYS.CTRL, KEYS.W], () => 'ctrl+W was pressed');
    return <Whatever />;
}
```

Reply:
It can receive an entry of KEYS or an array of KEYS :)